File: tools/tests/java/RoadCrewShadowTilesTest.java
package net.osmand.plus.roadcrew;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

// Test 106 (Galin, 2026-09-13): the phone asks for fixed tiles instead of its own
// 30 km box. The grid must be the server's (vectors from truck-map-tiles.ts), the
// 2x2 coverage must hold the phone at least half a tile from the edge, and the
// merge must follow the server's rules. Run with "keys lat,lon ..." to print keys
// for the cross-check against the server code in roadcrew-shadow-tiles.test.mjs.
public final class RoadCrewShadowTilesTest {

	public static void main(String[] args) {
		if (args.length > 0 && args[0].equals("keys")) {
			for (int i = 1; i < args.length; i++) {
				String[] point = args[i].split(",");
				System.out.println(RoadCrewShadowTiles.tileKey(Double.parseDouble(point[0]), Double.parseDouble(point[1])));
			}
			return;
		}
		serverVectors();
		coverageIsTheNearerTwoByTwo();
		coverageShrinksAtTheEdgesOfTheWorld();
		mergeKeepsABorderSegmentOnceAndTheCautiousVersion();
		mergeIsCompleteOnlyWhenEveryRequestedTileIsComplete();
		mergeDoesNotDependOnTileOrder();
		System.out.println("shadow tiles scenarios PASS");
	}

	private static void serverVectors() {
		// backend/roadcrew-api/src/truck-map-tiles.ts, tileKey(tileOf(lat, lon)), 13.09.
		String[][] vectors = {
				{"43.2141", "27.9147", "831_532"}, {"43.25", "27.75", "831_533"},
				{"43.2499999", "27.7499999", "830_532"}, {"0", "0", "720_360"},
				{"-0.0001", "-0.0001", "719_359"}, {"90", "180", "1439_719"},
				{"-90", "-180", "0_0"}, {"42.6977", "23.3219", "813_530"},
				{"44.4268", "26.1025", "824_537"}, {"-33.8688", "151.2093", "1324_224"},
				{"89.99", "-179.99", "0_719"}};
		for (String[] v : vectors) {
			String key = RoadCrewShadowTiles.tileKey(Double.parseDouble(v[0]), Double.parseDouble(v[1]));
			check(key.equals(v[2]), v[0] + "," + v[1] + " -> " + key + ", server says " + v[2]);
		}
		check(Arrays.equals(RoadCrewShadowTiles.bounds("831_532"), new double[] {43, 43.25, 27.75, 28}), "bounds of 831_532");
	}

	private static void coverageIsTheNearerTwoByTwo() {
		// Varna: in the east and north part of 831_532, so the east and north neighbours.
		check(RoadCrewShadowTiles.coverage(43.2141, 27.9147)
				.equals(Arrays.asList("831_532", "832_532", "831_533", "832_533")), "Varna coverage");
		// Sofia 42.6977, 23.3219 is in the north-west part of 813_530 (42.5-42.75, 23.25-23.5).
		check(RoadCrewShadowTiles.coverage(42.6977, 23.3219)
				.equals(Arrays.asList("812_530", "813_530", "812_531", "813_531")), "Sofia coverage");
		// The phone is at least half a tile from the edge of what it holds.
		for (double lat = 42.01; lat < 44; lat += 0.037) {
			for (double lon = 22.01; lon < 29; lon += 0.041) {
				double minLat = 90, maxLat = -90, minLon = 180, maxLon = -180;
				for (String key : RoadCrewShadowTiles.coverage(lat, lon)) {
					double[] b = RoadCrewShadowTiles.bounds(key);
					minLat = Math.min(minLat, b[0]); maxLat = Math.max(maxLat, b[1]);
					minLon = Math.min(minLon, b[2]); maxLon = Math.max(maxLon, b[3]);
				}
				double margin = Math.min(Math.min(lat - minLat, maxLat - lat), Math.min(lon - minLon, maxLon - lon));
				check(margin >= RoadCrewShadowTiles.TILE_SIZE_DEGREES / 2 - 1e-9, "margin at " + lat + "," + lon + " is " + margin);
			}
		}
	}

	private static void coverageShrinksAtTheEdgesOfTheWorld() {
		check(RoadCrewShadowTiles.coverage(89.99, -179.99).equals(Collections.singletonList("0_719")), "pole and antimeridian");
		check(RoadCrewShadowTiles.coverage(0.01, 179.99).size() == 2, "antimeridian: no wrap");
	}

	private static RoadCrewShadowTiles.Segment segment(String id, boolean eligible, double validUntil) {
		return new RoadCrewShadowTiles.Segment(id, eligible, validUntil,
				"{\"segmentId\":\"" + id + "\",\"eligible\":" + eligible + ",\"validUntil\":" + validUntil + "}");
	}

	private static void mergeKeepsABorderSegmentOnceAndTheCautiousVersion() {
		RoadCrewShadowTiles.Tile a = new RoadCrewShadowTiles.Tile("1_1", 2000, true,
				Arrays.asList(segment("border", true, 9000), segment("only-a", false, Double.POSITIVE_INFINITY)));
		RoadCrewShadowTiles.Tile b = new RoadCrewShadowTiles.Tile("2_1", 1000, true,
				Collections.singletonList(segment("border", false, Double.POSITIVE_INFINITY)));
		RoadCrewShadowTiles.Merge merged = RoadCrewShadowTiles.merge(Arrays.asList("1_1", "2_1"), Arrays.asList(a, b));
		check(merged.segments.size() == 2, "border segment kept once");
		for (RoadCrewShadowTiles.Segment s : merged.segments) {
			if (s.segmentId.equals("border")) check(!s.eligible, "not eligible wins over eligible");
		}
		check(merged.generatedAtMillis == 1000, "as old as the oldest tile");
		check(merged.complete, "both requested tiles complete");
		RoadCrewShadowTiles.Merge earlier = RoadCrewShadowTiles.merge(Collections.singletonList("1_1"), Arrays.asList(
				new RoadCrewShadowTiles.Tile("1_1", 1, true, Arrays.asList(segment("s", true, 9000), segment("s", true, 5000)))));
		check(earlier.segments.get(0).validUntil == 5000, "the earlier validity wins");
	}

	private static void mergeIsCompleteOnlyWhenEveryRequestedTileIsComplete() {
		RoadCrewShadowTiles.Tile full = new RoadCrewShadowTiles.Tile("1_1", 1, true, new ArrayList<>());
		RoadCrewShadowTiles.Tile cut = new RoadCrewShadowTiles.Tile("2_1", 1, false, new ArrayList<>());
		check(!RoadCrewShadowTiles.merge(Arrays.asList("1_1", "2_1"), Arrays.asList(full, cut)).complete, "a tile at its limit");
		check(!RoadCrewShadowTiles.merge(Arrays.asList("1_1", "2_1"), Arrays.asList(full, null)).complete, "a requested tile missing");
		check(RoadCrewShadowTiles.merge(Collections.singletonList("1_1"), Arrays.asList(full, cut)).complete, "a tile not requested is ignored");
		check(RoadCrewShadowTiles.merge(new ArrayList<>(), Arrays.asList(full)).generatedAtMillis == 0, "nothing requested is not an answer");
	}

	private static void mergeDoesNotDependOnTileOrder() {
		RoadCrewShadowTiles.Tile a = new RoadCrewShadowTiles.Tile("1_1", 1, true, Collections.singletonList(segment("s", true, 7000)));
		RoadCrewShadowTiles.Tile b = new RoadCrewShadowTiles.Tile("2_1", 1, true, Collections.singletonList(segment("s", true, 7000.5)));
		List<String> keys = Arrays.asList("1_1", "2_1");
		String one = RoadCrewShadowTiles.merge(keys, Arrays.asList(a, b)).segments.get(0).json;
		String two = RoadCrewShadowTiles.merge(keys, Arrays.asList(b, a)).segments.get(0).json;
		check(one.equals(two), "same result whatever the order");
	}

	private static void check(boolean condition, String message) {
		if (!condition) {
			throw new AssertionError(message);
		}
	}
}

File: tools/tests/roadcrew-shadow-tiles.test.mjs
import assert from 'node:assert/strict';
import { existsSync, readFileSync, readdirSync, mkdtempSync, rmSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join, resolve, dirname, basename } from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';
import { spawnSync } from 'node:child_process';
import test from 'node:test';

// Test 106 (Galin, 2026-09-13): the phone asks for fixed tiles
// (/v1/truck-map/shadow-tiles/) instead of its own 30 km box (shadow-segments).
// On 13.09 every box request read about 4,700 rows - the largest reader of the
// database; a tile is built once per 15-minute epoch and shared by every phone
// in the area.

const root = fileURLToPath(new URL('../../', import.meta.url));
const downloader = readFileSync(join(root,
  'OsmAnd/src/net/osmand/plus/roadcrew/RoadCrewShadowSnapshotDownloader.java'), 'utf8').replace(/\r\n/g, '\n');

function compile() {
  const tempRoot = resolve(tmpdir());
  const output = mkdtempSync(join(tempRoot, 'roadcrew-shadow-tiles-'));
  const result = spawnSync('javac', ['-d', output,
    join(root, 'OsmAnd/src/net/osmand/plus/roadcrew/RoadCrewShadowTiles.java'),
    join(root, 'tools/tests/java/RoadCrewShadowTilesTest.java')], { encoding: 'utf8', timeout: 30000 });
  assert.equal(result.status, 0, `${result.error ?? ''}\n${result.stderr}`);
  return {
    output,
    run: (args = []) => spawnSync('java', ['-cp', output, 'net.osmand.plus.roadcrew.RoadCrewShadowTilesTest', ...args],
      { encoding: 'utf8', timeout: 20000 }),
    cleanup: () => {
      assert.equal(dirname(resolve(output)), tempRoot);
      assert.ok(basename(output).startsWith('roadcrew-shadow-tiles-'));
      rmSync(output, { recursive: true, force: true });
    },
  };
}

// Test 107 (Galin, 2026-09-15): the phone talks to the project's own server
// directly; Cloudflare only forwards for older versions until they are gone.
test('every RoadCrew request goes to the own server, none to Cloudflare workers.dev', () => {
  const folder = join(root, 'OsmAnd/src/net/osmand/plus/roadcrew');
  const endpoints = readFileSync(join(folder, 'RoadCrewEndpoints.java'), 'utf8');
  assert.match(endpoints, /static final String API_BASE_URL = "https:\/\/api\.roadcrew\.meriltrans\.com";/);
  const offenders = readdirSync(folder).filter(file => file.endsWith('.java'))
    .filter(file => /workers\.dev|galin-b-vasilev1/.test(readFileSync(join(folder, file), 'utf8')));
  assert.deepEqual(offenders, []);
});

test('the own server is trusted on every Android version the app supports', () => {
  const manifest = readFileSync(join(root, 'OsmAnd/AndroidManifest.xml'), 'utf8');
  assert.match(manifest, /android:networkSecurityConfig="@xml\/roadcrew_network_security_config"/);
  // With a config present the manifest's cleartext flag is ignored; it must be kept here.
  assert.match(manifest, /android:usesCleartextTraffic="true"/);
  const config = readFileSync(join(root, 'OsmAnd/res/xml/roadcrew_network_security_config.xml'), 'utf8');
  assert.match(config, /<base-config cleartextTrafficPermitted="true">/);
  assert.match(config, /<domain includeSubdomains="false">api\.roadcrew\.meriltrans\.com<\/domain>/);
  for (const root_ of ['x1', 'x2']) {
    assert.match(config, new RegExp(`@raw/roadcrew_isrg_root_${root_}`));
    const pem = readFileSync(join(root, `OsmAnd/res/raw/roadcrew_isrg_root_${root_}.pem`), 'utf8');
    assert.match(pem, /-----BEGIN CERTIFICATE-----/);
  }
});

test('the downloader asks for tiles, never for its own box', () => {
  assert.match(downloader, /RoadCrewEndpoints\.API_BASE_URL \+ "\/v1\/truck-map\/shadow-tiles\/"/);
  assert.doesNotMatch(downloader, /shadow-segments/);
  assert.match(downloader, /RoadCrewShadowTiles\.coverage\(latitude, longitude\)/);
  assert.match(downloader, /RoadCrewShadowTiles\.merge\(coverage, parts\)/);
  assert.match(downloader, /now - fetched\.fetchedAtMillis >= RoadCrewShadowTiles\.TILE_EPOCH_MILLIS/,
    'a tile is fetched again only after its epoch');
  assert.match(downloader, /MAX_SNAPSHOT_SEGMENTS = 4 \* RoadCrewShadowTiles\.MAX_SEGMENTS_PER_TILE/);
});

// Real dependency-free production class, not a reimplementation.
test('shadow tiles runtime scenarios', () => {
  const java = compile();
  try {
    const run = java.run();
    assert.equal(run.status, 0, `${run.error ?? ''}\n${run.stdout}\n${run.stderr}`);
    assert.match(run.stdout, /shadow tiles scenarios PASS/);
  } finally {
    java.cleanup();
  }
});

// The phone's grid against the server's own code, when the backend is next to this repo.
const serverTiles = join(root, '..', 'backend', 'roadcrew-api', 'src', 'truck-map-tiles.ts');
test('the phone grid equals the server grid on 300 points', { skip: !existsSync(serverTiles) && 'backend not beside android' }, async () => {
  const { tileKey, tileOf } = await import(pathToFileURL(serverTiles).href);
  const points = [];
  let seed = 20260913;
  const next = () => (seed = (seed * 1103515245 + 12345) % 2147483648) / 2147483648;
  for (let i = 0; i < 250; i++) points.push([next() * 180 - 90, next() * 360 - 180]);
  for (let i = 0; i < 50; i++) points.push([Math.round((next() * 180 - 90) * 4) / 4, Math.round((next() * 360 - 180) * 4) / 4]);
  const java = compile();
  try {
    const run = java.run(['keys', ...points.map(([lat, lon]) => `${lat},${lon}`)]);
    assert.equal(run.status, 0, run.stderr);
    const phone = run.stdout.trim().split(/\r?\n/);
    const server = points.map(([lat, lon]) => tileKey(tileOf(lat, lon)));
    assert.deepEqual(phone, server);
  } finally {
    java.cleanup();
  }
});
